### ISSUE TITLE
zig fmt: better handling of struct with single field and complicated expression

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#1782]
```zig
pub fn init(out_stream: *Stream) Self
{
    return Self
    {
        .out_stream = switch(is_packed)
        {
            true => BitOutStream(endian, Stream.Error).init(out_stream),
            false => out_stream,
        },
    };
}
```

zig fmt does this:

```zig
pub fn init(out_stream: *Stream) Self {
    return Self{ .out_stream = switch (is_packed) {
        true => BitOutStream(endian, Stream.Error).init(out_stream),
        false => out_stream,
    } };
}
```

this looks ok to me, but the switch statement is used where an if is more canonical. with an if statement, zig fmt does this:

```zig
pub fn init(out_stream: *Stream) Self {
    return Self{ .out_stream = if (is_packed) BitOutStream(endian, Stream.Error).init(out_stream) else out_stream };
}
```

that's a bit excessive for one line. Ignoring a column limit, this is a complicated expression. 2 ways that would be reasonable to write, are not allowed by zig fmt:

```zig
// zig fmt forces the field expr init to be on the same line as `Self{`
pub fn init(out_stream: *Stream) Self {
    return Self{
        .out_stream = if (is_packed) BitOutStream(endian, Stream.Error).init(out_stream) else out_stream,
    };
}
```

or

```zig
pub fn init(out_stream: *Stream) Self {
    return Self{
        .out_stream = if (is_packed)
            BitOutStream(endian, Stream.Error).init(out_stream)
        else
            out_stream,
    };
}
```

zig fmt changes this to

```zig
pub fn init(out_stream: *Stream) Self {
    return Self{ .out_stream = if (is_packed)
        BitOutStream(endian, Stream.Error).init(out_stream)
    else
        out_stream };
}
```

which looks a bit awkward.

I don't have a specific proposal for how to fix this yet.

**Comment by hryx** [comment]
Maybe we need to define what's "too complicated" for a single line. The field value AST node can inform this somehow, either by its depth/size or its children's types.

We could restrict the single-line output for when the value is a single literal or identifier, but that doesn't seem fair to short expressions like this:

```zig
// allowed to occupy a single line
return Self{ .contents = first + second };
return Self{ .contents = newTortoise("Jeremy") };

// forced onto multiple lines
return Self{
    .contents = first + second,
};
return Self{
    .contents = newTortoise("Jeremy"),
};
```

How about this:

1. Crawl the tree starting at the field value expression node.
2. Increment a counter or "weight" value for each child node encountered. Maybe give more or less weight to certain node types?
3. If a threshold is reached, stop walking and decide on multiline. Otherwise, single-line.

I'm no expert, but if I recall, clang-format uses a weighting or "penalty" system sorta like this to decide when to insert line breaks in some contexts.

**Comment by andrewrk** [comment]
Let's start with some examples and decide what the results should be, and then we can see if this traversal idea would result in the desired behavior.

Two things to keep in mind:
 * `zig fmt` is not aware of line length by design. Policing column limits is not in scope of `zig fmt`.
 * The input is allowed to affect the output. While some things will be unconditionally reformatted, some things can be decided by "hints" from the input formatting.

Before we start doing something complicated, I think this issue might actually be solved quite simply, by allowing multiline struct literals with only 1 field:

 * When the input has a trailing comma (`,`), always do multiline struct literals.
 * When the input omits a trailing comma, always do single line struct literals.

There is precedent for this with function parameters, and I've been happy with the results. This would make both of the 2 reasonable ways in my comment above remain intact, and it would reformat the last "awkward" example to the desired version above it.

**Comment by hryx** [comment]
> - When the input has a trailing comma (`,`), always do multiline struct literals.
> - When the input omits a trailing comma, always do single line struct literals.

This sounds awesome. Consistent and way simpler than my idea. If line length is gonna be art-over-science, this sounds best.

I realize now the problem is not so much the _complexity_ of the field value expression, but the fact that if that expression gets rendered onto multiple lines, then the struct should be too. I think this would be the updated rules:

- If the input has a trailing comma **or any of the expressions outputs to multiple lines**, render a multiline struct literal.
- Otherwise, render a single-line struct literal.

That should prevent this from happening:

```zig
pub fn init(out_stream: *Stream) Self {
    // Because this if/else is multiline, that should have caused
    // the whole struct literal to be multiline
    return Self{ .out_stream = if (is_packed)
        BitOutStream(endian, Stream.Error).init(out_stream)
    else
        out_stream };
}
```

---

Based on those rules, these would all be valid outputs:

```zig
// one field
const a1 = A{ .x = false };
const a2 = A{
    .x = false,
};
const a3 = A{ .x = maybe() or backup() };
const a4 = A{
    .x = maybe() or backup(),
};

// multiple fields
const b1 = B{ .x = 1, .y = 2 };
const b2 = B{
    .x = 1,
    .y = 2,
};
const b3 = B{ .x = if (maybe()) 10 else 20, .y = operation() + 100 };
const b4 = B{
    .x = if (maybe()) 10 else 20,
    .y = operation() + 100,
};

// longer expression, value may be single or multi line
pub fn init1(out_stream: *Stream) Self {
    return Self{ .out_stream = if (is_packed) BitOutStream(endian, Stream.Error).init(out_stream) else out_stream };
}
pub fn init2(out_stream: *Stream) Self {
    return Self{
        .out_stream = if (is_packed) BitOutStream(endian, Stream.Error).init(out_stream) else out_stream,
    };
}
pub fn init3(out_stream: *Stream) Self {
    return Self{
        .out_stream = if (is_packed)
            BitOutStream(endian, Stream.Error).init(out_stream)
        else
            out_stream,
    };
}
```

**Comment by andrewrk** [comment]
Good observation - I agree on all points.
